File: contractsdk/cpp/example/builtin_types.cc
#include <cstdio>
#include <cinttypes>
#include "xchain/xchain.h"


struct BuiltinTypes : public xchain::Contract {};

void print_tx(xchain::Transaction t);
void print_block(xchain::Block b);

DEFINE_METHOD(BuiltinTypes, initialize) {
    xchain::Context* ctx = self.context();
    ctx->ok("initialize succeed");
}

DEFINE_METHOD(BuiltinTypes, gettx) {
    xchain::Context* ctx = self.context();
    const std::string& txid = ctx->arg("txid");
    xchain::Transaction tx;
     
    if (ctx->query_tx(txid, &tx)) {
        print_tx(tx);
        ctx->ok("Obtaining transaction succeed");
    } else {
        ctx->error("txid not found");
    }
}

DEFINE_METHOD(BuiltinTypes, getblock) {
    xchain::Context* ctx = self.context();
    const std::string& blockid = ctx->arg("blockid");
    xchain::Block block;

    if (ctx->query_block(blockid, &block)) {
        print_block(block);
        ctx->ok("Obtaining block succeed");
    } else {
        ctx->error("blockid not found");
    }
}

void print_block(xchain::Block b) {
    printf("[Block]:\n");
    printf("blockid: %s\n", b.blockid.c_str());    
    printf("pre_hash: %s\n", b.pre_hash.c_str());    
    printf("proposer: %s\n", b.proposer.c_str());    
    printf("sign: %s\n", b.sign.c_str());    
    printf("pubkey: %s\n", b.pubkey.c_str());    
    printf("height: %" PRId64"\n", b.height);    
    for (auto v : b.txids) {
        printf("txid: %s\n", v.c_str());
    }
    printf("tx_count: %d\n", b.tx_count);    
    printf("in_trunk: %d\n", b.in_trunk);    
    printf("next_hash: %s\n", b.next_hash.c_str());    
}

void print_tx(xchain::Transaction t) {
    printf("[Transaction]:\n");
    printf("txid: %s\n", t.txid.c_str());    
    printf("blockid: %s\n", t.blockid.c_str());    
    printf("desc: %s\n", t.desc.c_str());    
    printf("initiator: %s\n", t.initiator.c_str());    
    for (auto v : t.auth_require) {
        printf("auth_require: %s\n", v.c_str());
    }
    for (int i = 0; i < t.tx_inputs.size(); i++) {
        printf("[tx_input[%d]]: ref_txid: %s\n", i, t.tx_inputs[i].ref_txid.c_str());
        printf("[tx_input[%d]]: ref_offset: %d\n", i, t.tx_inputs[i].ref_offset);
        printf("[tx_input[%d]]: from_addr: %s\n", i, t.tx_inputs[i].from_addr.c_str());
        printf("[tx_input[%d]]: amount: %s\n", i, t.tx_inputs[i].amount.c_str());
    }
    for (int i = 0; i < t.tx_outputs.size(); i++) {
        printf("[tx_output[%d]]: ammount: %s\n", i, t.tx_outputs[i].amount.c_str());
        printf("[tx_output[%d]]: to_addr: %s\n", i, t.tx_outputs[i].to_addr.c_str());
    }
}

File: core/contractsdk/java/example/builtin-types/src/main/java/com/baidu/xuper/BuiltinTypes.java
package com.baidu.xuper;

import com.baidu.xuper.contractpb.Contract.Block;
import com.baidu.xuper.contractpb.Contract.Transaction;
import com.google.protobuf.ProtocolStringList;

import java.io.UnsupportedEncodingException;
import java.math.BigInteger;
import java.util.Iterator;
import java.util.List;

/**
 * Builtin Types
 */
public class BuiltinTypes implements Contract {

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        return Response.ok("initialize success".getBytes());
    }

    @ContractMethod
    public Response getTx(Context ctx) {
        byte[] txidByte = ctx.args().get("txid");
        if (txidByte == null) {
            return Response.error("missing txid");
        }
        String txid = new String(txidByte);

        Transaction tx = ctx.queryTx(txid);
        printTx(ctx, tx);

        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response getBlock(Context ctx) {
        byte[] blockidByte = ctx.args().get("blockid");
        if (blockidByte == null) {
            return Response.error("missing blockid");
        }
        String blockid = new String(blockidByte);

        Block b = ctx.queryBlock(blockid);
        printBlock(ctx, b);

        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response authRequire(Context ctx) {

        List<String> authRequireList = ctx.authRequire();
        printAuthRequire(ctx, authRequireList);

        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    // transfer native token from the contract to other accounts
    public Response transfer(Context ctx) {
        byte[] toByte = ctx.args().get("to");
        if (toByte == null) {
            return Response.error("missing to");
        }
        String to = new String(toByte);

        byte[] amountByte = ctx.args().get("amount");
        if (amountByte == null) {
            return Response.error("missing amount");
        }
        String amountStr = new String(amountByte);
        BigInteger amount = new BigInteger(amountStr);
        if (amount.signum() == -1) {
            return Response.error("amount must not be negative");
        }

        ctx.transfer(to, amount);

        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response put(Context ctx) {
        byte[] key = ctx.args().get("key");
        if (key == null) {
            return Response.error("missing key");
        }

        byte[] value = ctx.args().get("value");
        if (value == null) {
            return Response.error("missing value");
        }

        ctx.putObject(key, value);
        return Response.ok(value);
    }

    @ContractMethod
    public Response get(Context ctx) {
        byte[] key = ctx.args().get("key");
        if (key == null) {
            return Response.error("missing key");
        }

        byte[] value = ctx.getObject(key);
        if (value == null) {
            return Response.error("key " + new String(key) + " not found");
        }

        return Response.ok(value);
    }

    @ContractMethod
    public Response getList(Context ctx) {
        byte[] start = ctx.args().get("start");
        if (start == null) {
            return Response.error("missing start");
        }

        byte[] limit = PrefixRange.generateLimit(start);
        Iterator<ContractIteratorItem> iter = ctx.newIterator(start, limit);
        int i = 0;
        while (iter.hasNext()) {
            ContractIteratorItem item = iter.next();
            String key = bytesToString(item.getKey());
            String value = bytesToString(item.getValue());
            ctx.log("item: " + i + ", key: " + key + ", value: " + value);
            i++;
        }

        return Response.ok("ok".getBytes());
    }

    private void printTx(Context ctx, Transaction tx) {
        ctx.log("txid: " + tx.getTxid());
        ctx.log("blockid: " + tx.getBlockid());
        ctx.log("blockid: " + tx.getDesc().toStringUtf8());
        ctx.log("initiator: " + tx.getInitiator());

        ProtocolStringList authRequireList = tx.getAuthRequireList();
        for (String auth : authRequireList) {
            ctx.log("auth require: " + auth);
        }

        for (int i = 0; i < tx.getTxInputsCount(); i++) {
            ctx.log("tx_input: " + i + ", ref_txid: " + tx.getTxInputs(i).getRefTxid());
            ctx.log("tx_input: " + i + ", ref_offset: " + tx.getTxInputs(i).getRefOffset());
            ctx.log("tx_input: " + i + ", from_addr: " + tx.getTxInputs(i).getFromAddr().toStringUtf8());
            ctx.log("tx_input: " + i + ", amount: " + tx.getTxInputs(i).getAmount());
        }

        for (int i = 0; i < tx.getTxOutputsCount(); i++) {
            ctx.log("tx_output: " + i + ", amount: " + tx.getTxOutputs(i).getAmount());
            ctx.log("tx_output: " + i + ", to_addr: " + tx.getTxOutputs(i).getToAddr().toStringUtf8());
        }
    }

    private void printBlock(Context ctx, Block b) {
        ctx.log("blockid: " + b.getBlockid());
        ctx.log("pre_hash: " + b.getPreHash());
        ctx.log("proposer: " + b.getProposer().toStringUtf8());
        ctx.log("sign: " + b.getSign());
        ctx.log("pubkey: " + b.getPubkey().toStringUtf8());
        ctx.log("height: " + b.getHeight());

        for (int i = 0; i < b.getTxCount(); i++) {
            ctx.log("txid: " + i + ", " + b.getTxids(i));
        }

        ctx.log("tx_count: " + b.getTxCount());
        ctx.log("in_trunk: " + b.getInTrunk());
        ctx.log("next_hash: " + b.getNextHash());
    }

    private void printAuthRequire(Context ctx, List<String> authRequireList) {
        for (int i = 0; i < authRequireList.size(); i++) {
            ctx.log("authRequire: " + i + ", " + authRequireList.get(i));
        }
    }

    private static String bytesToString(byte[] bytes) {
        if (null == bytes || bytes.length == 0) {
            return "";
        }
        String strContent = "";
        try {
            strContent = new String(bytes, "utf-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return strContent;
    }

    public static void main(String[] args) {
        Driver.serve(new BuiltinTypes());
    }
}
